models/repo_indexer.go

```diff
@@ -0,0 +1,294 @@
+		Data: &indexer.RepoIndexerData{
+			RepoID: repo.ID,
+		},
+	})
+}
+
+// genesisChanges get changes to add repo to the indexer for the first time
+func genesisChanges(repo *Repository) (*repoChanges, error) {
+	var changes repoChanges
+	stdout, err := git.NewCommand("ls-files").RunInDir(repo.LocalCopyPath())
+	if err != nil {
+		return nil, err
+	}
+	for _, line := range strings.Split(stdout, "\n") {
+		filename := strings.TrimSpace(line)
+		if len(filename) == 0 {
+			continue
+		} else if filename[0] == '"' {
+			filename, err = strconv.Unquote(filename)
+			if err != nil {
+				return nil, err
+			}
+		}
+		changes.UpdatedFiles = append(changes.UpdatedFiles, filename)
+	}
+	return &changes, nil
+}
+
+// nonGenesisChanges get changes since the previous indexer update
+func nonGenesisChanges(repo *Repository) (*repoChanges, error) {
+	diffCmd := git.NewCommand("diff", "--name-status",
+		repo.IndexerStatus.CommitSha, "HEAD")
+	stdout, err := diffCmd.RunInDir(repo.LocalCopyPath())
+	if err != nil {
+		// previous commit sha may have been removed by a force push, so
+		// try rebuilding from scratch
+		if err = indexer.DeleteRepoFromIndexer(repo.ID); err != nil {
+			return nil, err
+		}
+		return genesisChanges(repo)
+	}
+	var changes repoChanges
+	for _, line := range strings.Split(stdout, "\n") {
+		line = strings.TrimSpace(line)
+		if len(line) == 0 {
+			continue
+		}
+		filename := strings.TrimSpace(line[1:])
+		if len(filename) == 0 {
+			continue
+		} else if filename[0] == '"' {
+			filename, err = strconv.Unquote(filename)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		switch status := line[0]; status {
+		case 'M', 'A':
+			changes.UpdatedFiles = append(changes.UpdatedFiles, filename)
+		case 'D':
+			changes.RemovedFiles = append(changes.RemovedFiles, filename)
+		default:
+			log.Warn("Unrecognized status: %c (line=%s)", status, line)
+		}
+	}
+	return &changes, nil
+}
+
+func updateLastIndexSync(repo *Repository) error {
+	stdout, err := git.NewCommand("rev-parse", "HEAD").RunInDir(repo.LocalCopyPath())
+	if err != nil {
+		return err
+	}
+	sha := strings.TrimSpace(stdout)
+	return repo.updateIndexerStatus(sha)
+}
+
+func processRepoIndexerOperationQueue() {
+	for {
+		op := <-repoIndexerOperationQueue
+		if op.deleted {
+			if err := indexer.DeleteRepoFromIndexer(op.repo.ID); err != nil {
+				log.Error(4, "DeleteRepoFromIndexer: %v", err)
+			}
+		} else {
+			if err := updateRepoIndexer(op.repo); err != nil {
+				log.Error(4, "updateRepoIndexer: %v", err)
+			}
+		}
+	}
+}
+
+// DeleteRepoFromIndexer remove all of a repository's entries from the indexer
+func DeleteRepoFromIndexer(repo *Repository) {
+	addToOperationQueue(repoIndexerOperation{repo: repo, deleted: true})
+}
+
+// UpdateRepoIndexer update a repository's entries in the indexer
+func UpdateRepoIndexer(repo *Repository) {
+	addToOperationQueue(repoIndexerOperation{repo: repo, deleted: false})
+}
+
+func addToOperationQueue(op repoIndexerOperation) {
```

Comment:
What does this function do?

Reply:
It adds `op` to `repoIndexerOperationQueue` (tries to add without spawning a new thread, but spawns a new thread if adding to queue will block).

Reply:
Maybe rename to `addOperationToQueue`

Reply:
Done